lcm/src/lcm/providers/udpm.rs

```diff
@@ -71,21 +73,25 @@ pub struct UdpmProvider<'a> {
 }
 impl<'a> UdpmProvider<'a> {
     /// Creates a new UDPM provider using the given settings.
-    pub fn new(network: &str, options: &HashMap<&str, &str>) -> Result<Self, InitError> {
+    pub fn new(url: &Url) -> Result<Self, InitError> {
         // Parse the network string into the address and port
-        let (addr, port) = UdpmProvider::parse_network_string(network)?;
-
-        // Get the TTL value
-        let ttl = match options.get("ttl").unwrap_or(&"0").parse() {
-            Ok(ttl) => ttl,
-            Err(_) => return Err(InitError::InvalidLcmUrl),
-        };
+        let addr = url.to_socket_addrs()?.next().expect("The URL should contain an address");
```

Comment:
I have two concerns about this, but both are out of ignorance:
* How do we know that the URL has a valid `SocketAddr`? I didn't search very hard, but I couldn't find anything in the `url` crate that says it will have one and the documentation for `ToSocketAddr` says that it is valid to yield zero.
* How do we know that the first one is the correct one? The code for things like `UdpSocket` try every value the iterator yields before giving up.

But it seems like there's a good chance that this will be rewritten before release so it's probably not worth worrying about.

Reply:
The documentation for `ToSocketAddrs` describes it as:
>A trait for objects which can be converted or resolved to one or more SocketAddr values.

Based on that description, I'd consider it a bug if an implementor returned an iterator of zero elements, since they should return an error type instead in that case. The docs for `Url` say:
>/// Return an error if `Url::host` or `Url::port_or_known_default` return `None`.

and the code looks like it does what it says it does.

Reply:
The trait returns an iterator because things like DNS queries can return multiple addresses. Most types that implement `ToSocketAddrs` are going to return just one.

I would be very surprised if any users of LCM are using an address that requires a DNS lookup.

Reply:
The description of the `to_socket_addrs()` method is where I'm reading that zero is a valid number to return:

> The returned iterator may not actually yield any values depending on the outcome of any resolution performed.

Doesn't resolution from `/etc/hosts` count as DNS resolution? Network programming is really not my thing, so apologies if I sound ignorant, but I could definitely see people using something in `/etc/hosts` as the address.

Reply:
That's a good point - `/etc/hosts` does count. I was thinking more along the lines of domains like `google.com`, which may resolve to many different IPs. I don't think anyone expects LCM to work with that URL!

The docs seem to contradict themselves, but the implementation of `Url` always returns one element or an error.

Reply:
If it doesn't, I'm sure someone will make an issue and we can fix it then.